File: lib/bloc/dog_bloc.dart
import 'package:flutter_bloc/flutter_bloc.dart';
import '../repositories/dog_repository.dart';
import 'dog_event.dart';
import 'dog_state.dart';

class DogBloc extends Bloc<DogEvent, DogState> {
  final DogRepository repository;

  DogBloc({required this.repository}) : super(DogInitial()) {
    on<FetchDogImage>(_onFetchDogImage);
  }

  Future<void> _onFetchDogImage(
    FetchDogImage event,
    Emitter<DogState> emit,
  ) async {
    emit(DogLoading());
    try {
      final dogImage = await repository.getRandomDogImage();
      emit(DogLoaded(dogImage));
    } catch (e) {
      emit(DogError(e.toString()));
    }
  }
} 

File: app/src/main/java/com/example/dogapp/bloc/DogBloc.kt
package com.example.dogapp.bloc

import com.example.dogapp.repository.DogRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class DogBloc(private val repository: DogRepository) {
    private val _state = MutableStateFlow<DogState>(DogState.Initial)
    val state: StateFlow<DogState> = _state

    fun onEvent(event: DogEvent) {
        when (event) {
            is DogEvent.FetchDogImage -> fetchDogImage()
        }
    }

    private fun fetchDogImage() {
        CoroutineScope(Dispatchers.IO).launch {
            _state.value = DogState.Loading
            try {
                val dogImage = repository.getRandomDogImage()
                _state.value = DogState.Loaded(dogImage)
            } catch (e: Exception) {
                _state.value = DogState.Error(e.message ?: "Unknown error")
            }
        }
    }
} 